From 9d2ca569945fde2a67f608cbe2e448acc69df66c Mon Sep 17 00:00:00 2001
From: Pradeep P V K <pradeep.pragallapati@oss.qualcomm.com>
Date: Mon, 17 Aug 2026 15:58:31 +0530
Subject: [PATCH] tee_supplicant: Allow qtee_supplicant to access the eMMC RPMB
 device

Add SELinux allow rules for qtee_supplicant to access the eMMC RPMB
device, enabling communication with the RPMB partition on eMMC-based
systems.

Define a new storage_raw_rw_removable_device() interface to storage.if to allow
read or write on removable chr_file interfaces, and use it in
tee_supplicant.te to grant qtee_supplicant the necessary chr_file
rw_chr_file_perms on removable_device_t.

Upstream-Status: Backport [https://github.com/SELinuxProject/refpolicy/commit/9d2ca569945fde2a67f608cbe2e448acc69df66c]
Signed-off-by: Pradeep P V K <pradeep.pragallapati@oss.qualcomm.com>
---
 policy/modules/kernel/storage.if          | 18 ++++++++++++++++++
 policy/modules/services/tee_supplicant.te |  3 +++
 2 files changed, 21 insertions(+)

diff --git a/policy/modules/kernel/storage.if b/policy/modules/kernel/storage.if
index 08ca90a1c5..2165b6f13f 100644
--- a/policy/modules/kernel/storage.if
+++ b/policy/modules/kernel/storage.if
@@ -910,6 +910,24 @@ interface(`storage_dontaudit_raw_write_removable_device',`
 	dontaudit $1 removable_device_t:blk_file write_blk_file_perms;
 ')
 
+########################################
+## <summary>
+#	read or write on removable device interfaces.
+## </summary>
+## <param name="domain">
+##     <summary>
+##     Domain allowed access.
+##     </summary>
+## </param>
+#
+interface(`storage_raw_rw_removable_device',`
+	gen_require(`
+		type removable_device_t;
+	')
+
+	allow $1 removable_device_t:chr_file rw_chr_file_perms;
+')
+
 ########################################
 ## <summary>
 ##	Allow the caller to directly read
diff --git a/policy/modules/services/tee_supplicant.te b/policy/modules/services/tee_supplicant.te
index ab0cc2e8c3..b0cf3f5413 100644
--- a/policy/modules/services/tee_supplicant.te
+++ b/policy/modules/services/tee_supplicant.te
@@ -63,4 +63,7 @@ tunable_policy(`tee_supplicant_qtee',`
 
         # Access qtee_supplicant to send logs to systemd journal
         logging_send_syslog_msg(tee_supplicant_t)
+
+        # Allow qtee_supplicant to access the eMMC RPMB device (/dev/mmcblk0rpmb)
+        storage_raw_rw_removable_device(tee_supplicant_t)
 ')
